src/net/sourceforge/kolmafia/textui/langserver/Script.java

```diff
@@ -0,0 +1,193 @@
+
+/**
+ * A file that was recognized as a KoLmafia script in one of the directories under our authority.
+ */
+public class Script {
+  final AshLanguageServer parent;
+  final File file;
+
+  /**
+   * The {@link Handler} in charge of this script. <i>Only</i> exists if this script is <b>not</b>
+   * imported by any other script.
+   *
+   * <p>When trying to find the handler(s) in charge of a given file, {@link
+   * FilesMonitor#findHandlers(File)} should be used.
+   */
+  Handler handler;
+
+  Script(final AshLanguageServer parent, final File file) {
+    this.parent = parent;
+    this.file = file;
+  }
+
+  Handler makeHandler() {
+    this.handler = new Handler();
+
+    this.parent.executor.execute(
+        () -> {
+          this.handler.parseFile();
+        });
+
+    return this.handler;
+  }
+
+  /**
+   * An object tasked with taking care of a {@link Script}, such as parsing it and sending the
+   * resulting diagnostics to the client once the parsing ends.
+   *
+   * <p>All files imported by this script should also be handled by this object
+   */
+  public class Handler {
+    Parser parser;
+    Scope scope;
+
+    private Thread parserThread;
+
+    private final Object parserSwapLock = new Object();
+    private final Object parserThreadWaitingLock = new Object();
+
+    // Public class, private constructor
+    private Handler() {}
+
+    private void parseFile() {
+      final String previousThreadName = Thread.currentThread().getName();
+
+      synchronized (this.parserSwapLock) {
+        this.parserThread = Thread.currentThread();
+        this.parserThread.setName(Script.this.file.getName() + " - Parser");
+      }
+
+      this.parser =
+          new LSParser(Script.this.file, null, Collections.synchronizedMap(new HashMap<>()));
+
+      try {
+        // Parse the script
+        this.scope = this.parser.parse();
+
+        // If we managed to parse it without interruption, send the diagnostics
+        Script.this.parent.executor.execute(
+            () -> {
+              this.sendDiagnostics();
+            });
+      } catch (InterruptedException e) {
+      } finally {
+        synchronized (this.parserSwapLock) {
+          if (this.parserThread == Thread.currentThread()) {
+            this.parserThread = null;
+
+            synchronized (this.parserThreadWaitingLock) {
+              this.parserThreadWaitingLock.notifyAll();
+            }
+          }
+
+          Thread.currentThread().setName(previousThreadName);
+        }
+      }
+    }
+
+    void sendDiagnostics() {
+      if (Script.this.parent.client == null) {
+        return;
+      }
+
+      this.waitForParsing();
+
+      if (Script.this.handler != this || this.parser == null || Thread.interrupted()) {
+        // We've been kicked out
+        return;
+      }
+
+      final String previousThreadName = Thread.currentThread().getName();
+      // Will technically send the diagnostics of every file it imports, but
```

Comment:
What is "it"? Does this thread send the diagnostics of the assigned file (`Script.this.file`), plus the diagnostics of every file that file imports?

Reply:
It does. That's the point of Handler. (i.e. to be in charge of a File, plus every script it imports)